File: portal/backend/common/pagination.py
# -*- coding: utf-8 -*-
"""分页参数解析工具"""

from flask import request

# 单页最多返回行数，防止 ?per_page=1000000 整表拉取
MAX_PER_PAGE = 200
DEFAULT_PER_PAGE = 20


def parse_pagination():
    """从请求参数解析并钳制 page / per_page，返回 (page, per_page)。"""
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', DEFAULT_PER_PAGE, type=int)
    page = max(page, 1)
    per_page = min(max(per_page, 1), MAX_PER_PAGE)
    return page, per_page

File: apps/api/src/common/pagination.ts
/**
 * Pagination parameter parsing
 */

/** Max rows per page; prevents pulling a whole table via ?per_page=1000000 */
export const MAX_PER_PAGE = 200
export const DEFAULT_PER_PAGE = 20

/** Parse a query param as int; falls back to the default when missing or unparsable */
export function queryInt(value: unknown, fallback: number): number {
  const raw = Array.isArray(value) ? value[0] : value
  if (typeof raw !== 'string') return fallback
  const text = raw.trim().replace(/_/g, '')
  if (!/^[+-]?\d+$/.test(text)) return fallback
  return Number.parseInt(text, 10)
}

/** Parse and clamp page / per_page from the query */
export function parsePagination(query: Record<string, unknown> = {}): { page: number; per_page: number } {
  const page = Math.max(queryInt(query.page, 1), 1)
  const perPage = Math.min(Math.max(queryInt(query.per_page, DEFAULT_PER_PAGE), 1), MAX_PER_PAGE)
  return { page, per_page: perPage }
}
